### PR TITLE
fix(automl): fix data leakage in classification holdout validation

### PR DESCRIPTION
**Pull request by commint-tian**
>>
>> - Stop duplicating X_first in both train/val sets
>> - Implement dynamic class balancing
>> - Preserve original dataset size

## Why are these changes needed?

In generic_task.py, the unconditional merging of X_first into both the training and validation sets leads to data duplication, causing the total row count len(X_train) + len(X_val) to exceed the original dataset size. More critically, this introduces a data leakage issue: since part of the validation data (X_first) also appears in the training set, the evaluation results may become unreliable.

To resolve this, we can dynamically check for missing classes using missing_in_train and missing_in_val to identify gaps in the training/validation sets, then selectively supplement data:

Validation set only: Add X_first (ensuring validation set class coverage takes priority).

Training set: Fill missing classes by sampling from the remaining data X_rest (avoiding reuse of X_first).

## Related issue number
issue #1390

## Checks

- [ ] I've used [pre-commit](https://microsoft.github.io/FLAML/docs/Contribute#pre-commit) to lint the changes in this PR (note the same in integrated in our CI checks).
- [ ] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by commint-tian** [comment]
@microsoft-github-policy-service agree

File: flaml/automl/task/generic_task.py [hunk 1]
```diff
@@ -514,45 +514,40 @@ def prepare_data(
...
                 X_train, X_val, y_train, y_val = self._train_test_split(
                     state, X_rest, y_rest, first, rest, split_ratio, stratify
                 )
-                X_train = concat(X_first, X_train)
-                y_train = concat(label_set, y_train) if data_is_df else np.concatenate([label_set, y_train])
-                X_val = concat(X_first, X_val)
-                y_val = concat(label_set, y_val) if data_is_df else np.concatenate([label_set, y_val])
+                #Check whether the training set and validation set cover all categories.
+                train_labels = np.unique(y_train)
+                val_labels = np.unique(y_val)
```
Comment (thinkall) [inline flaml/automl/task/generic_task.py:537]: np.unique doesn't work for psSeries or psDataFrame. Check an example [here](https://github.com/microsoft/FLAML/blob/d131f97c7bfd436c6a53d68d43451934e3da9f92/flaml/automl/spark/utils.py#L136).

File: flaml/automl/task/generic_task.py [hunk 2]
```diff
@@ -514,45 +514,40 @@ def prepare_data(
...
-                        if data_is_df
-                        else X_train_all[rest]
-                    )
-                    y_rest = (
-                        iloc_pandas_on_spark(y_train_all, rest)
-                        if is_spark_dataframe
-                        else y_train_all.iloc[rest]
-                        if data_is_df
-                        else y_train_all[rest]
-                    )
```
[the remark this answers is not in the document]
Comment (commint-tian) [reply flaml/automl/task/generic_task.py:547]: Oh，I made a mistake about it. The second part should be kept here.

File: flaml/automl/task/generic_task.py (hunk 1 above)
Comment (commint-tian) [reply flaml/automl/task/generic_task.py:537]: it can be modified like this：
if isinstance(y_train, (ps.Series, ps.DataFrame)):
    train_labels = y_train.unique() if isinstance(y_train, ps.Series) else y_train.iloc[:, 0].unique()
else:
    train_labels = np.unique(y_train)

if isinstance(y_val, (ps.Series, ps.DataFrame)):
    val_labels = y_val.unique() if isinstance(y_val, ps.Series) else y_val.iloc[:, 0].unique()
else:
    val_labels = np.unique(y_val)

File: flaml/automl/task/generic_task.py [hunk 3]
```diff
@@ -514,45 +514,40 @@ def prepare_data(
...
+                #Add X_first only to the validation set (remove the merge of the training set).
+                if missing_in_val:
+                    X_val = concat(X_first, X_val)
+                    y_val = concat(label_set, y_val) if data_is_df else np.concatenate([label_set, y_val])
+                #The training set supplements the missing categories with the remaining data.
+                if missing_in_train:
+                    for label in missing_in_train:
+                        mask = (y_rest == label)
+                        X_train = concat(X_rest[mask], X_train)
+                        y_train = concat(y_rest[mask], y_train)
```
[the remark this answers is not in the document]
Comment (commint-tian) [reply flaml/automl/task/generic_task.py:550]: Perhaps this could work：
mask = (y_rest == label)
filtered_X_rest = X_rest.filter(mask)
filtered_y_rest = y_rest.filter(mask)

X_train = X_train.union(filtered_X_rest)
y_train = y_train.union(filtered_y_rest)

File: flaml/automl/task/generic_task.py (hunk 1 above)
Comment (thinkall) [reply flaml/automl/task/generic_task.py:537]: try reusing existing functions

File: flaml/automl/task/generic_task.py [hunk 4]
```diff
@@ -514,45 +514,40 @@ def prepare_data(
...
+                val_labels = np.unique(y_val)
+                missing_in_train = set(label_set) - set(train_labels)
+                missing_in_val = set(label_set) - set(val_labels)
+
+                #Add X_first only to the validation set (remove the merge of the training set).
+                if missing_in_val:
+                    X_val = concat(X_first, X_val)
+                    y_val = concat(label_set, y_val) if data_is_df else np.concatenate([label_set, y_val])
+                #The training set supplements the missing categories with the remaining data.
+                if missing_in_train:
```
[the remark this answers is not in the document]
Comment (thinkall) [reply flaml/automl/task/generic_task.py:546]: a better way is to fill the missing labels in both train and val with first. For those not missing in either train or val, put them in train.

File: flaml/automl/task/generic_task.py (hunk 1 above)
Comment (commint-tian) [reply flaml/automl/task/generic_task.py:537]: this is a pretty big project, and I'm not very familiar with it yet. I'm not sure if there are already existing functions that I can use.

File: flaml/automl/task/generic_task.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (commint-tian) [reply flaml/automl/task/generic_task.py:546]: Indeed, I've learned it.

File: flaml/automl/task/generic_task.py (hunk 1 above)
Comment (thinkall) [reply flaml/automl/task/generic_task.py:537]: > Check an example [here](https://github.com/microsoft/FLAML/blob/d131f97c7bfd436c6a53d68d43451934e3da9f92/flaml/automl/spark/utils.py#L136).

Check the example [here](https://github.com/microsoft/FLAML/blob/d131f97c7bfd436c6a53d68d43451934e3da9f92/flaml/automl/spark/utils.py#L136).

**Comment by thinkall** [comment]
Hi @commint-tian , do you get a chance to revise the PR? Thanks.

File: flaml/automl/task/generic_task.py (hunk 1 above)
Comment (commint-tian) [reply flaml/automl/task/generic_task.py:537]: This is the modified code：
train_labels= len_labels(y_train)
val_labels= len_labels(y_val)

File: flaml/automl/task/generic_task.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (commint-tian) [reply flaml/automl/task/generic_task.py:546]: if missing_in_val:
    if data_is_df:
        X_val = concat([X_first, X_val])
        y_val = concat([label_set, y_val])
    else:
        X_val = np.concatenate([X_first, X_val], axis=0)
        y_val = np.concatenate([label_set, y_val])

if missing_in_train:
    if data_is_df:
        X_train = concat([X_first, X_train])
        y_train = concat([label_set, y_train])
    else:
        X_train = np.concatenate([X_first, X_train], axis=0)
        y_train = np.concatenate([label_set, y_train])

common_labels = set(train_labels) & set(val_labels)
only_train_labels = set(train_labels) - set(val_labels)
only_val_labels = set(val_labels) - set(train_labels)

# 将不在train或val中缺失的标签放到train中
for label in only_val_labels:
    # 从val中移除这些标签
    mask = y_val != label
    X_val = X_val[mask]
    y_val = y_val[mask]

    # 将这些标签添加到train中
    if data_is_df:
        X_train = concat([X_train, X_first.loc[X_first[label_col] == label]])
        y_train = concat([y_train, label_set.loc[label_set[label_col] == label]])
    else:
        X_train = np.concatenate([X_train, X_first[y_first == label]], axis=0)
        y_train = np.concatenate([y_train, label_set[y_first == label]])

**Comment by commint-tian** [comment]
> Hi @commint-tian , do you get a chance to revise the PR? Thanks.

I've finished revising the code, please check it again.

**Comment by thinkall** [comment]
> > Hi @commint-tian , do you get a chance to revise the PR? Thanks.
> 
> I've finished revising the code, please check it again.

Hi @commint-tian , have you pushed your changes to github? I don't see it yet.
